### ISSUE TITLE
eval runner: CE/perplexity over a TOYC/packed corpus — the consuming half of the F9 eval seam (toy#129 item 3)

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#130]
toy#129 item 3 split in two. The ENABLING half shipped in bed358c: `toy train franken --ckpt-every N` writes standard fused-llama checkpoints at boundaries (CPU + CUDA), gated sched-null (curve byte-equals the no-ckpt run). This issue is the CONSUMING half: the CE evaluator those checkpoints feed.

**Why offline/checkpoint-boundary instead of in-process** (the design decision, recorded): an in-process eval forward on the training session collides with the process-global sched (the one-graph-twins rule) and the F1.1 one-sched-alloc-on-graph_b constraint; the lr=0 trick pollutes Adam moments. Checkpoint-boundary evals sidestep all three, and quality-per-FLOP joins on `step` across bundles.

**Ask** — a new spinel unit (landmine #16), `toy eval-ce` or similar:
1. Load a fused-llama GGUF (the franken checkpoints qualify; `toy eval`'s mmap path is the precedent — today's `toy eval` is the frozen-point logprob slice and explicitly defers CE-over-corpus).
2. Stream a pack (TOYC header → vocab authoritative; headerless + `--vocab`), forward whole ctx-length windows through the seq engine (build_forward + forward-only compute — no backward, no params), and compute mean CE / perplexity over `--eval-tokens` M tokens from a caller-chosen offset (held-out discipline = disjoint slice/pack, Tao-side).
3. Deterministic output lines (`eval_ce: windows=<n> tokens=<n> ce=<v> ppl=<v>`) + a toy/v1 `eval` event bundle (own events.jsonl per eval run; joining into the training bundle would break the run_end-last invariant — keep them separate files).
4. Gate: determinism 2×, a known-model sanity band, TOYC/headerless both.

**Perf note from the item-1 measurements**: at vocab 50257 the one-hot label path costs ctx×vocab floats per window (12.8M at ctx 256 — already a visible constant in training). For eval, prefer downloading logits and computing CE Ruby-side from int targets, or an int-target CE node if the engine grows one — do NOT inherit the one-hot build.

F9 runs its quality-per-FLOP read as: train with `--ckpt-every N` → eval each checkpoint on the held-out slice → join eval CE against run_start.cost (toy#129 item 4) and step wall_us.

**Comment by OriPekelman** [comment]
Delivered in f959ddf — the evaluator, the CLI, AND the MoE lane's read (a gap the ticket didn't cover), full battery green at 41 legs.

## `toy eval ce --ckpt <gguf> --corpus <pack> [--context N] [--eval-tokens M] [--offset K] [--out dir] [--json]`

New `libexec/toy-eval-ce` unit riding eval_lmc's gated machinery single-checkpoint: random-init session → overwrite every param from the fused GGUF (per-head attention slices) → lr=0 AdamW windows → **ggml-internal CE** (byte-exact, the lmc determinism discipline). Design points:

- **lr=0 is clean here**: the moments it mutates belong to a throwaway session — the toy#122 pollution objection was about the *training* session.
- **The one-hot perf note resolved differently than the ticket sketched**: keeping CE in ggml beats downloading logits (a Ruby-side log-softmax over ctx×vocab per window would be slower than the upload); the one-hot Mat is built once and only the previous T scatter positions are cleared per window — no ctx×vocab refill. The ~50 MB/window upload is dominated by the forward.
- **Eval never repeats data**: no rotation; EOF truncates loudly, zero windows fails.
- **Vocab discipline**: model vocab is the authority — TOYC-header disagreement and any OOB token id fail loud (gated).

## The MoE lane (not in the ticket, load-bearing for F9)

The demonstrator IS franken-moe — which has no GGUF writer, so checkpoint-boundary evals can't carry it. The MoE runner now evals **in-runner after training** (`toy train franken-moe ... --eval-corpus P --eval-tokens M --eval-offset K`): lr=0 windows on the live graph, sound because the Adam moments are dead once training ends. The training curve is gated **byte-null** against a no-eval run; the `eval` event lands before `run_end`.

## The whole F9 loop, smoked

200-step franken run on the GPT-2 smoke pack with `--ckpt-every 50`, each checkpoint evaled at a held-out offset: **CE ladder 9.91 → 8.76 → 8.81 → 8.69** (train loss 8.13 — held-out sits above it, as it must). MoE wide/E=4/bp-spine: held-out 8.89 vs train 7.77.

F9 is fully runnable: fixture (toy#129 item 1) + cost accounting (item 4) + no-shadow (item 2) + `--ckpt-every` → `toy eval ce` for llama arms, `--eval-corpus` for MoE arms, joined on step × run_start.cost × wall_us.

**Comment by OriPekelman** [comment]
Closing: shipped in `f959ddf` — feat(eval): toy eval ce + MoE end-of-run held-out eval — the F9 quality read (toy#130)

Verified by `git log`; the implementing commit names this issue in its subject. Part of the 2026-08-14 issue-state sweep (48 closed, all against a commit in `main`). Reopen if the scope here outran what landed.
